conda_recipe_manager/fetcher/git_artifact_fetcher.py

```diff
@@ -4,40 +4,120 @@
 
 from __future__ import annotations
 
+import shutil
 from pathlib import Path
+from typing import Final, NamedTuple, Optional
+
+from git import Repo
+from git import exc as git_exceptions
 
 from conda_recipe_manager.fetcher.base_artifact_fetcher import BaseArtifactFetcher
+from conda_recipe_manager.fetcher.exceptions import FetchError
+
+
+class _GitTarget(NamedTuple):
+    """
+    Convenience structure that contains all the information to target a `git` repo at a point in time.
+    """
+
+    url: str
+    branch: Optional[str] = None
+    tag: Optional[str] = None
+    rev: Optional[str] = None
 
 
 class GitArtifactFetcher(BaseArtifactFetcher):
     """
     Artifact Fetcher capable of cloning a remote git repository.
     """
 
-    def __init__(self, name: str, git_url: str):
+    def __init__(
+        self, name: str, url: str, branch: Optional[str] = None, tag: Optional[str] = None, rev: Optional[str] = None
+    ):
         """
-        TODO
-        TODO add other params
+        Constructs a `GitArtifactFetcher` instance.
+
+        :param name: Identifies the artifact. Ideally, this is the package name. In multi-sourced/mirrored scenarios,
+            this might be the package name combined with some identifying information.
+        :param url: Remote or local path to the target repository
+        :param branch: (Optional) Target git branch name
+        :param tag: (Optional) Target git tag name
+        :param rev: (Optional) Target git revision ID
         """
         super().__init__(name)
-        self._git_url = git_url
 
-    def _clone(self) -> None:
+        self._is_remote = not Path(url).exists()
+        self._git_target = _GitTarget(
+            url=url,
+            branch=branch,
+            tag=tag,
+            rev=rev,
+        )
+
+    def _clone(self) -> Repo:
         """
-        TODO
+        Helper function that clones a git repository from a local or remote source.
+
+        :raises GitError: If there was an issue cloning the target git repository.
+        :raises IOError: If there is an issue with the file system.
         """
-        pass
+        match self._is_remote:
+            case True:
+                return Repo.clone_from(self._git_target.url, self._temp_dir_path)  # type: ignore[misc]
+            case False:
+                shutil.copytree(self._git_target.url, self._temp_dir_path)
+                return Repo(self._temp_dir_path)
+
+    def _resolve_checkout_target(self) -> Optional[str]:
```

Comment:
`Optional[str]` can be spelled `str | None`

Reply:
I know, but this is arguably more human readable and in-line with nomenclature used by other languages .

Reply:
well, no... it boils down to preference other than human-readable.
I can argue that ``str | None`` is more readable since, I know the two types of the return when you compare to ``Optional[str]``, well, what is ``Optional`` in python? that is more implicit.

and which programming languages? typescript uses ``str | null``, rust you can use ``Option<T>`` but usually put that in a result.
swift and C# uses ``str?``

Reply:
I have spent a significant amount of time with modern C++ and that uses [optional](https://en.cppreference.com/w/cpp/utility/optional). Maybe I'm biased from that. Apparently [Java does as well](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html) and I would argue `rust`'s `Option` is just two letters away from being `Option`-`al`.

Regardless, I don't find style arguments worth anyone's time. There are much bigger problems to solve in our community.

This is the lane we picked in this project and I choose to stay in that lane unless I hear someone claim there is a significant practical difference.